x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/mapper/CartesianPoint.java

```diff
@@ -0,0 +1,239 @@
+                + "but found: [{}]", vals.length);
+        }
+        final float x;
+        final float y;
+        try {
+            x = Float.parseFloat(vals[0].trim());
+            if (Float.isFinite(x) == false) {
+                throw new ElasticsearchParseException("invalid x value [{}]; " +
+                    "must be between -3.4028234663852886E38 and 3.4028234663852886E38", x);
+            }
+         } catch (NumberFormatException ex) {
+            throw new ElasticsearchParseException("x must be a number");
+        }
+        try {
+            y = Float.parseFloat(vals[1].trim());
+            if (Float.isFinite(y) == false) {
+                throw new ElasticsearchParseException("invalid y value [{}]; " +
+                    "must be between -3.4028234663852886E38 and 3.4028234663852886E38", y);
+            }
+        } catch (NumberFormatException ex) {
+            throw new ElasticsearchParseException("y must be a number");
+        }
+        return reset(x, y);
+    }
+
+    private CartesianPoint resetFromWKT(String value) {
+        Geometry geometry;
+        try {
+            geometry = new WellKnownText(false, new StandardValidator(true))
+                .fromWKT(value);
+        } catch (Exception e) {
+            throw new ElasticsearchParseException("Invalid WKT format", e);
+        }
+        if (geometry.type() != ShapeType.POINT) {
+            throw new ElasticsearchParseException("[geo_point] supports only POINT among WKT primitives, " +
+                "but found " + geometry.type());
+        }
+        org.elasticsearch.geometry.Point point = (org.elasticsearch.geometry.Point) geometry;
+        return reset((float) point.getX(), (float) point.getY());
+    }
+
+    public float getX() {
+        return this.x;
+    }
+
+    public float getY() {
+        return this.y;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        CartesianPoint point = (CartesianPoint) o;
+
+        if (Float.compare(point.x, x) != 0) return false;
+        if (Float.compare(point.y, y) != 0) return false;
+
+        return true;
+    }
+
+    @Override
+    public int hashCode() {
+        int result;
+        int temp;
+        temp = x != +0.0f ? Float.floatToIntBits(x) : 0;
+        result = Integer.hashCode(temp);
+        temp = y != +0.0f ? Float.floatToIntBits(y) : 0;
+        result = 31 * result + Integer.hashCode(temp);
+        return result;
+    }
+
+    @Override
+    public String toString() {
+        return x + ", " + y;
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        return builder.startObject().field("x", x).field("y", y).endObject();
+    }
+
+    public static CartesianPoint parsePoint(XContentParser parser, CartesianPoint point)
```

Comment:
I know this is largely a mirror of `GeoPoint` but I'm wondering if we can achieve this using `ObjectParser` or `ConstructingObjectParser`? (something we've been meaning to address w/ `GeoPoint` as well). I'm not sure how that works with the parsing leniency we have on Point and GeoPoint (e.g., array vs comma delimited string).  Curious if @nik9000 or @talevy have some ideas. Maybe that can be done in a followup PR.